src/serac/numerics/petsc_solvers.cpp

```diff
@@ -0,0 +1,1031 @@
+// Copyright (c) 2019-2024, Lawrence Livermore National Security, LLC and
+// other Serac Project Developers. See the top-level LICENSE file for
+// details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#include "serac/numerics/petsc_solvers.hpp"
+#include "serac/infrastructure/logger.hpp"
+
+#ifdef SERAC_USE_PETSC
+
+#include "petsc/private/matimpl.h"
+#include "petsc/private/linesearchimpl.h"
+#include "petscmathypre.h"
+
+namespace serac::mfem_ext {
+
+// Static functions needed to create a shell PC
+
+/// @brief Context to store wrapped solver object
+typedef struct {
+  /// @brief Wrapped solver object
+  mfem::Solver* solver;
+
+  /// @brief Flag indicating whether @a solver is owned by the preconditioner, generally false.
+  bool owns_op;
+} SolverWrapperCtx;
+
+/**
+ * @brief Callback for printing information about the preconditioner object to a viewer
+ *
+ * @param[in] pc Shell preconditioner object to print
+ * @param[out] viewer Viewer to output information to
+ *
+ * @return Error code, or 0 on success.
+ */
+static PetscErrorCode solverWrapperView(PC pc, PetscViewer viewer)
+{
+  SolverWrapperCtx* ctx;
+
+  PetscFunctionBeginUser;
+  auto* void_ctx = static_cast<void*>(&ctx);
+  PetscCall(PCShellGetContext(pc, &void_ctx));
+  ctx = static_cast<SolverWrapperCtx*>(void_ctx);
+  if (ctx->solver) {
+    mfem::PetscPreconditioner* ppc = dynamic_cast<mfem::PetscPreconditioner*>(ctx->solver);
```

Comment:
Stream of conciousness..... so MFEM does "Petsc" and it looks like below this PETSc itself does Petsc???